### DISCUSSION TITLE
nanoseconds time stamps auto rounding off to microseconds in grafana - elasticsearch queries

### DISCUSSION THREAD

**Discussion by vmanju123** [grafana/grafana#61173] [Data Sources]
**What happened**:
we have a field in our ES database called nanoTS, long integer basically a number representing nano seconds Eg: 1586266796413334123, when we query this in grafana, it automatically rounds it to nearest microseconds Eg: 1586266796413334000, the same query in Kibana provides accurate number as is in database, we are using logs based query to get entire index into a table panel

**What you expected to happen**:
No rounding off or if there is a option to turn off rounding

**How to reproduce it (as minimally and precisely as possible)**:
add a field in es db for long number give numbers as provided above

**Anything else we need to know?**:

**Environment**:
- Grafana version: 6.5.2
- Data source type & version: ElastisSearch 7.0+
- OS Grafana is installed on: Ubuntu 18.04
- User OS & Browser: Windows/Chrome
- Grafana plugins: Out of box plugins
- Others:

**Comment by aocenas** [comment]
Hey @vmanju123 issue is JS by default does not handle big integers and there is no special handling of that right now. I assume it is your primary timestamp but if not you could workaround this by using it as string which at least would give you accurate display and sorting.

**Comment by vmanju123** [comment]
Its not our primary time stamp field, but additional field, when you say string, you meant on es database change the field from long to string ?

**Comment by aocenas** [comment]
Yes unfortunately that would have to be changed in elastic. The conversion to number in Grafana and thus loss of precision happens during the response parsing, so it's not possible to fix that with formatting later on.

**Comment by Estrax** [comment]
@aocenas do we need any additional work there, or can we close this issue?

**Comment by gabor** [comment]
@kylebrandt do you know maybe, is there a plan in Grafana to have a way to store very large integers (larger than `2^^53-1`) in a dataframe-field? because currently numbers-in-a-field are represented as javascript-numbers in the javascript parts of Grafana, and those numbers break down above `2^^53-1` ( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER ).

**Comment by kylebrandt** [comment]
@gabor I am not sure - @leeoniya or @ryantxu would likely be the best people to ask. We moved from Arrow encoding to JSON for performance reasons, so the tricky part is handling the data on the frontend. In the backend Arrow is still the encoding over the wire for plugin datasources to Grafana - so the JSON encoding is for FE (JSON API) requests.

**Comment by gabor** [comment]
yeah, i assume in the backend-code this all works just fine (large intergers should be OK in `go`... i think)..

**Comment by leeoniya** [comment]
i think we should encode these as ms timestamp, plus ns or us offsets in the field values. we can use our planned encoding optimization for repeated large values as well:

https://github.com/grafana/grafana/blob/dbb33eaba90cd8b0edacf8d5c3e2b4f09e3d7f92/packages/grafana-data/src/dataframe/DataFrameJSON.ts#L44

in terms of displaying nanosecond values, we'll probably need many additional UI component changes to expect/handle ms+offset tuples...or something. it's gonna be a while till Temporal apis are widely available: https://github.com/tc39/proposal-temporal

**Comment by leeoniya** [comment]
hmm, actually this will only let us pass these values through the json dataframe. but they'll get decoded automatically into field values where this would break down again. we'll need to figure out a way to skip this auto-decoding into full field values. tricky :(

we definitely have plans to support for some kind of base + offset[] thing. it's useful for additional usecases beyond the integer limit situation.

**Comment by gabor** [comment]
@leeoniya thanks for the feedback. but, i think in this specific case, they are not considered timestamps by grafana, it is just a large integer-value (i mean, these values are not in the "timestamp" field of the dataframe, there are just a numeric-field in elasticsearch)

**Comment by leeoniya** [comment]
so there are two issues here. the first is the limit on integer size. we can detect this and encode as needed to avoid precision loss.

the second is the intent is to treat them as timestamps. i dont see how to square that circle without giving users explicit settings in the datasource UI to set specific numeric field's `type` as `time` or `time_ns`. but i'm just edu guessing here.

**Comment by gabor** [comment]
> so there are two issues here. the first is the limit on integer size. we can detect this and encode as needed to avoid precision loss.

could you describe this part in more detail?

> the second is the intent is to treat them as timestamps. i dont see how to square that circle without giving users explicit settings in the datasource UI to set specific numeric field's `type` as `time` or `time_ns`. but i'm just edu guessing here.

i think we can ignore this problem. i mean, the original issue here is about showing large-integers without deforming those large-integers (because of javascript.MAX_SAFE_INT). there is no need to do anything "timestamp-related", afaik.

**Comment by suiluj** [comment]
is this an ongoing discussion?

as far as i know the ordering problem is not fixed because of this.

I have the same problem with nanosecond precision logs in grafana

another issue describes the wrong ordering problem for nanosecond logs stored in loki: https://github.com/grafana/grafana/issues/23457

**Reply by leeoniya** [reply]
recent update: https://github.com/grafana/grafana/issues/6252#issuecomment-1566460589

**Reply by suiluj** [reply]
Ah thanks for the hint. :)

**Reply by leeoniya** [reply]
i guess specifically for logs/ordering, you'll want to watch https://github.com/grafana/grafana/issues/67889

**Reply by gabor** [reply]
hi @suiluj , could you describe your problem in more detail? some things work correctly already. so for example, if you use Loki, the timestamps should be sorted nanosecond-correctly.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). This looks to be solved. If something is still missing, feel free to make an [issue](https://github.com/grafana/grafana/issues) if you cannot find a solution and would like to request the feature.
